### ISSUE TITLE
Hot reload button not working

### ISSUE THREAD

**Issue by pzoltowski** [flutter/website#9230]
### Page URL

https://docs.flutter.dev/get-started/test-drive/

### Page source

https://github.com/flutter/website/tree/main/src/get-started/test-drive/index.md

### Describe the problem

Hot Reload (yellow lightning) button in VSCode doesn't work. I get in console:
`Reloaded 0 libraries in 91ms (compile: 7 ms, reload: 0 ms, reassemble: 47 ms).`

but text label is still old one. On the other hand doing just Cmd + S (save) *do* works though - so overall it's not a big deal but worth to investigate.  When pressing Cmd+S I get correct reload message in console:

`Reloaded 1 of 665 libraries in 156ms (compile: 12 ms, reload: 76 ms, reassemble: 40 ms).`

### Expected fix

investigate either why VSCode hot reload button doesn't  work or update documentation that only Save All works.

### Additional context

Macbook 16'' M2 Max
MacOS Ventura 13.4.1 (c) 
Xcode 14.3.1 (14E300c)
iPhone 13 mini with iOS 16.6
VSCode Version: 1.81.0

other console messages:
```
Launching lib/main.dart on [**REDACTED**] in debug mode...
main.dart:1
Automatically signing iOS for device deployment using specified development team in Xcode project: [**REDACTED**]
Xcode build done.                                           21.4s
(lldb) 2023-08-10 17:17:33.296154+0200 Runner[68843:5005374] [VERBOSE-2:FlutterDarwinContextMetalImpeller.mm(35)] Using the Impeller rendering backend.
Warning: Unable to create restoration in progress marker file
fopen failed for data file: errno = 2 (No such file or directory)
Errors found! Invalidating cache...
Connecting to VM Service at ws://127.0.0.1:52744/[**REDACTED**]=/ws
```

**Comment by parlough** [comment]
Hi @pzoltowski, thanks for opening an issue!

I'm not too familiar with VS Code, but I believe it might have auto save off by default. If you haven't saved your changes yet, hot reload won't be able to update your app with the new values. Saving works because it saves your file then runs the hot reload.

Could you try making sure auto save is enabled? VS Code has some documentation on how to do so at https://code.visualstudio.com/docs/editor/codebasics#_save-auto-save. 

You can also configure the automatic hot reload on save behavior with the [`dart.flutterHotReloadOnSave` configuration option](https://dartcode.org/docs/settings/#dartflutterhotreloadonsave).

Let us know how that goes. Thanks again!

**Comment by pzoltowski** [comment]
Thanks @parlough . That's correct I had  VSCode auto save off and flutter hot reload on save: manual (default one). 

I think on this default configuration VSCode hot reload button should either:

- save current file and hot reload (because clicking button sounds like doing this manual)

- or if feasible in VSCode: even if file is not saved (explicitly on disk) and dirty in VSCode still use this dirty version that is send to hot reload

Another possible things to improve:

1) flutter hot reload on save options (manual, all) seems only make difference when vscode autosave is on - otherwise they seems to behave the same and hard to figure out at first glance what's the difference.

2) hot reload button should change status (**enabled**/**disabled**) depending if there is anything to reload. Right now when autosave is **off** and flutter hot reload on save is **manual** because of point 1) when user not explicitly hit cmd+s hot reload seems that doesn't work (BTW option descriptions could probably be better what 'automatic' in 'all' means) . On the other hand when user hit cmd+s and app just got hot reloaded probably no point to have 'hot reload' button enabled right now because we just got hot reloaded. Unless there is any other corner case I'm not aware off?

**Comment by danagbemava-nc** [comment]
@DanTup can you please weigh-in here? I'm not sure if there's any action to be taken from the docs team on this

**Comment by pzoltowski** [comment]
@danagbemava-nc That's true that this might be actually more an issue to confirm with team behind VSCode plugin. For the time being I would just update doc from:

`Save your changes: invoke Save All, **or** click Hot Reload lightning bolt .`

to

Save your changes: invoke Save All, **and** click Hot Reload lightning bolt .

because clicking hot reload is never enough no matter what settings someone set. On the other hand with default settings Save All is just enough to trigger hot reload.

**Comment by DanTup** [comment]
Do I understand correctly that the issue here is that clicking the Hot Reload (lightning button) appears to not do anything if you have unsaved changes, because only the saved files are being hot-reloaded?

I don't know if we are able to save files when clicking this button (because the click is sent directly to the debug adapter, not to the extension code). It's possible we could disable it if there are unsaved files, although I don't know if that would be annoying if you're modifying something like a text file. I suspect few people actually use this button now hot-reload-on-save is the default though, so maybe tweaking the documentation to note that your files must be saved first would suffice (and perhaps noting that by default the act of saving will trigger a reload, so this button isn't usually required unless you have turned off hot-reload-on-save)?

**Comment by pzoltowski** [comment]
>Do I understand correctly that the issue here is that clicking the Hot Reload (lightning button) appears to not do anything if you have unsaved changes, because only the saved files are being hot-reloaded?

That's correct and in tutorial ('**or**') gave impression 'save all' is not mandatory. Long term this is not a big deal but probably better update tutorial doc so that someone who is doing flutter onboarding (like me) stay on track and not get distracted trying to figure out why it doesn't work.

And that's right it seems 'hot reload' button is only useful in someone change default vscode settings 'flutter hot reload on save' to 'never'

**Comment by DanTup** [comment]
> That's correct and in tutorial ('or') gave impression 'save all' is not mandatory.

I'm not certain, but I have a feeling this might have worked that way in the past. I'm wondering if VS Code previously was saving all files whenever a contributed debug toolbar button was clicked, and maybe it became clear that wasn't appropriate for all buttons. I'll see if I can find any info on this, because if it was changed perhaps we could as for a flag that would allow us to restore this behaviour for Hot Reload.

In the meantime I think it makes sense to tweak the docs here. We can always tweak them back if we get a way to automatically trigger a save.

**Comment by DanTup** [comment]
Actually, I just reviewed the code and I was mistaken - the debug toolbar clicks do not go directly go to the debugger, they go to the extension.

So, I suggest that we update the VS Code extension to automatically save all files whenever you click the hot reload button. If this might cause issues, we can add a setting to opt-out, and just have it on by default.

I've opened https://github.com/Dart-Code/Dart-Code/issues/4685 to track this. I suggest we don't bother changing the docs :-)
